File: go/internal/conversation/conversation.go
package conversation

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/lloydzhou/bash-agent/internal/protocol"
)

type Store struct {
	Path string
}

func (s Store) Ensure() error {
	f, err := os.OpenFile(s.Path, os.O_CREATE, 0o644)
	if err != nil {
		return err
	}
	return f.Close()
}

func (s Store) AddUser(content string) error {
	line, err := json.Marshal(map[string]any{
		"role":    "user",
		"content": content,
	})
	if err != nil {
		return err
	}
	return appendLine(s.Path, line)
}

func (s Store) AddAssistant(text string, thinking string, calls []protocol.ToolCallEvent) error {
	content := make([]any, 0, len(calls)+2)
	// Always emit thinking block (matches bash build_assistant_content_json)
	content = append(content, map[string]any{
		"type":     "thinking",
		"thinking": thinking,
	})
	// Always emit text block (matches bash build_assistant_content_json)
	content = append(content, map[string]any{
		"type": "text",
		"text": text,
	})
	for _, call := range calls {
		content = append(content, map[string]any{
			"type":  "tool_use",
			"id":    call.ID,
			"name":  call.Name,
			"input": json.RawMessage(call.InputJSON),
		})
	}
	line, err := json.Marshal(map[string]any{
		"role":    "assistant",
		"content": content,
	})
	if err != nil {
		return err
	}
	return appendLine(s.Path, line)
}

func (s Store) AddToolResults(results []ToolResult) error {
	content := make([]any, 0, len(results))
	for _, result := range results {
		convContent := result.ConvContent
		if convContent == "" {
			convContent = result.Content
		}
		content = append(content, map[string]any{
			"type":        "tool_result",
			"tool_use_id": result.ToolUseID,
			"content":     convContent,
		})
	}
	line, err := json.Marshal(map[string]any{
		"role":    "user",
		"content": content,
	})
	if err != nil {
		return err
	}
	return appendLine(s.Path, line)
}

func (s Store) MessagesJSON() ([]byte, error) {
	data, err := os.ReadFile(s.Path)
	if err != nil {
		return nil, err
	}
	var out bytes.Buffer
	out.WriteByte('[')
	first := true
	for _, line := range strings.Split(string(data), "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		if !first {
			out.WriteByte(',')
		}
		first = false
		out.WriteString(line)
	}
	out.WriteByte(']')
	return out.Bytes(), nil
}

func (s Store) Lines() ([]json.RawMessage, error) {
	data, err := os.ReadFile(s.Path)
	if err != nil {
		return nil, err
	}
	lines := make([]json.RawMessage, 0)
	for _, line := range strings.Split(string(data), "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		lines = append(lines, json.RawMessage(line))
	}
	return lines, nil
}

func (s Store) TrimKeepLast(keepLines int) error {
	if keepLines < 0 {
		keepLines = 0
	}
	lines, err := s.Lines()
	if err != nil {
		return err
	}
	if keepLines >= len(lines) {
		return nil
	}
	var out bytes.Buffer
	for _, line := range lines[len(lines)-keepLines:] {
		out.Write(line)
		out.WriteByte('\n')
	}
	return os.WriteFile(s.Path, out.Bytes(), 0o644)
}

func (s Store) TotalBytes() (int, error) {
	data, err := os.ReadFile(s.Path)
	if err != nil {
		return 0, err
	}
	return len(data), nil
}

func (s Store) TotalLines() (int, error) {
	lines, err := s.Lines()
	if err != nil {
		return 0, err
	}
	return len(lines), nil
}

// CountUserInputs counts the number of user_input events in events.jsonl
func (s Store) CountUserInputs() (int, error) {
	// Read events file (same path but with events.jsonl)
	eventsPath := strings.Replace(s.Path, "conversation.jsonl", "events.jsonl", 1)
	data, err := os.ReadFile(eventsPath)
	if err != nil {
		return 0, err
	}
	count := 0
	for _, line := range strings.Split(string(data), "\n") {
		if strings.Contains(line, `"type":"user_input"`) {
			count++
		}
	}
	return count, nil
}

func (s Store) KeepLineCount(targetBytes int) (int, error) {
	lines, err := s.Lines()
	if err != nil {
		return 0, err
	}
	if len(lines) == 0 {
		return 0, nil
	}
	sizes := make([]int, len(lines))
	turnStart := make([]bool, len(lines))
	for i, line := range lines {
		sizes[i] = len(line) + 1
		var msg struct {
			Role    string          `json:"role"`
			Content json.RawMessage `json:"content"`
		}
		if err := json.Unmarshal(line, &msg); err == nil {
			turnStart[i] = msg.Role == "user" && len(msg.Content) > 0 && msg.Content[0] == '"'
		}
	}
	keep := 0
	bytes := 0
	for i := len(lines) - 1; i >= 0; i-- {
		if keep > 0 && bytes+sizes[i] > targetBytes {
			break
		}
		bytes += sizes[i]
		keep++
	}
	if keep == 0 {
		keep = 1
	}
	start := len(lines) - keep
	adjusted := start
	for adjusted < len(lines) && !turnStart[adjusted] {
		adjusted++
	}
	if adjusted < len(lines) {
		start = adjusted
	} else {
		for start > 0 && !turnStart[start] {
			start--
		}
	}
	return len(lines) - start, nil
}

func appendLine(path string, line []byte) error {
	f, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return err
	}
	defer f.Close()
	if _, err := f.Write(append(line, '\n')); err != nil {
		return err
	}
	return nil
}

type ToolResult struct {
	ToolUseID   string
	ToolName    string
	ToolArgs    map[string]string
	Content     string // Full output for events.jsonl TOOL_RESULT (matches bash: write_message("TOOL_RESULT", id, name, output))
	ConvContent string // Summary only for conv file (matches bash: result_for_conv = first line of Edit output)
}

func BuildToolResultJSON(id, content string) ([]byte, error) {
	return json.Marshal(map[string]any{
		"type":        "tool_result",
		"tool_use_id": id,
		"content":     content,
	})
}

func FirstLine(s string) string {
	if i := strings.IndexByte(s, '\n'); i >= 0 {
		return s[:i]
	}
	return s
}

func BuildReadToolResultSummary(path string) string {
	if path == "" {
		return "Read"
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return fmt.Sprintf("Read(%s)", path)
	}
	return fmt.Sprintf("Read(%s) [%d lines, %d bytes]", path, lineCount(string(data)), len(data))
}

func BuildFileToolResultSummary(kind, path string) string {
	if path == "" {
		return kind
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return fmt.Sprintf("%s(%s)", kind, path)
	}
	return fmt.Sprintf("%s(%s) [%d lines, %d bytes]", kind, path, lineCount(string(data)), len(data))
}

func BuildEditDiffSummary(path, diff string) string {
	added, removed := 0, 0
	for _, line := range strings.Split(diff, "\n") {
		if strings.HasPrefix(line, "+") && !strings.HasPrefix(line, "+++") {
			added++
		}
		if strings.HasPrefix(line, "-") && !strings.HasPrefix(line, "---") {
			removed++
		}
	}
	return fmt.Sprintf("Edit(%s) [+%d -%d lines]", path, added, removed)
}

func BuildEditToolResultSummary(path, oldString, newString string) string {
	oldPreview := previewLines(oldString, 6)
	newPreview := previewLines(newString, 6)
	var b strings.Builder
	b.WriteString(fmt.Sprintf("Edit(%s)\n", path))
	if oldPreview == "" {
		b.WriteString("- (empty)\n")
	} else {
		for _, line := range strings.Split(oldPreview, "\n") {
			b.WriteString("- ")
			b.WriteString(line)
			b.WriteByte('\n')
		}
	}
	if newPreview == "" {
		b.WriteString("+ (empty)")
	} else {
		for _, line := range strings.Split(newPreview, "\n") {
			b.WriteString("+ ")
			b.WriteString(line)
			b.WriteByte('\n')
		}
	}
	return strings.TrimRight(b.String(), "\n")
}

	// ... tool result

func BuildToolCallSummary(name string, fields map[string]string) string {
	var label string
	switch name {
	case "Read", "Write", "Edit":
		label = fields["path"]
	case "Glob", "Grep":
		label = fields["pattern"]
	case "Bash":
		label = strings.ReplaceAll(fields["command"], "\n", " ")
		if len(label) > 80 {
			label = label[:77] + "..."
		}
	case "TodoWrite":
		label = fields["summary"]
	case "Skill":
		label = fields["name"]
	}
	if label == "" {
		return name
	}
	return fmt.Sprintf("%s(%s)", name, label)
}

func lineCount(s string) int {
	if s == "" {
		return 0
	}
	n := 1
	for _, ch := range s {
		if ch == '\n' {
			n++
		}
	}
	return n
}

func previewLines(s string, limit int) string {
	if s == "" {
		return ""
	}
	lines := strings.Split(s, "\n")
	if len(lines) > limit {
		lines = lines[:limit]
	}
	return strings.Join(lines, "\n")
}

File: rust/src/conversation.rs
use crate::protocol::ToolCallEvent;
use anyhow::Result;
use serde_json::{Value, json};
use std::fs;
use std::io::Write;
use std::path::PathBuf;

#[derive(Debug, Clone)]
pub struct Store {
    pub path: PathBuf,
}

#[derive(Debug, Clone)]
pub struct ToolResult {
    pub tool_use_id: String,
    pub tool_name: String,
    pub tool_args: std::collections::BTreeMap<String, String>,
    pub content: String, // Full output for events.jsonl TOOL_RESULT (matches bash)
    pub conv_content: String, // Summary only for conv file (matches bash: result_for_conv = first line of Edit output)
}

impl Store {
    pub fn ensure(&self) -> Result<()> {
        if !self.path.exists() {
            fs::File::create(&self.path)?;
        }
        Ok(())
    }

    pub fn add_user(&self, content: &str) -> Result<()> {
        self.append_line(&json!({"role":"user","content":content}))
    }

    pub fn add_assistant(&self, text: &str, thinking: &str, calls: &[ToolCallEvent]) -> Result<()> {
        let mut content = Vec::new();
        // Always emit thinking block (matches bash build_assistant_content_json)
        content.push(json!({"type":"thinking","thinking":thinking}));
        // Always emit text block (matches bash build_assistant_content_json)
        content.push(json!({"type":"text","text":text}));
        for c in calls {
            content.push(json!({"type":"tool_use","id":c.id,"name":c.name,"input":c.input_json}));
        }
        self.append_line(&json!({"role":"assistant","content":content}))
    }

    pub fn add_tool_results(&self, results: &[ToolResult]) -> Result<()> {
        let content: Vec<Value> = results
            .iter()
            .map(|r| {
                let conv_content = if r.conv_content.is_empty() {
                    &r.content
                } else {
                    &r.conv_content
                };
                json!({"type":"tool_result","tool_use_id":r.tool_use_id,"content":conv_content})
            })
            .collect();
        self.append_line(&json!({"role":"user","content":content}))
    }

    pub fn file_tool_result_summary(kind: &str, path: &str) -> String {
        if path.is_empty() {
            return kind.to_string();
        }
        match fs::read(path) {
            Ok(data) => format!(
                "{kind}({path}) [{} lines, {} bytes]",
                line_count(&data),
                data.len()
            ),
            Err(_) => format!("{kind}({path})"),
        }
    }

    pub fn lines(&self) -> Result<Vec<Value>> {
        let data = fs::read_to_string(&self.path)?;
        Ok(data
            .lines()
            .filter(|l| !l.trim().is_empty())
            .map(serde_json::from_str)
            .collect::<std::result::Result<Vec<Value>, _>>()?)
    }

    pub fn messages_json(&self) -> Result<Value> {
        Ok(Value::Array(self.lines()?))
    }

    pub fn total_bytes(&self) -> Result<usize> {
        Ok(fs::read(&self.path)?.len())
    }

    pub fn total_lines(&self) -> Result<usize> {
        Ok(self.lines()?.len())
    }

    /// Count user_input events in events.jsonl
    pub fn count_user_inputs(&self) -> Result<usize> {
        let events_path = self
            .path
            .to_string_lossy()
            .replace("conversation.jsonl", "events.jsonl");
        let data = fs::read_to_string(&events_path)?;
        let count = data
            .lines()
            .filter(|line| line.contains("\"type\":\"user_input\""))
            .count();
        Ok(count)
    }

    pub fn trim_keep_last(&self, keep_lines: usize) -> Result<()> {
        let data = fs::read_to_string(&self.path)?;
        let raw_lines: Vec<&str> = data.lines().filter(|l| !l.trim().is_empty()).collect();
        if keep_lines >= raw_lines.len() {
            return Ok(());
        }
        let kept = &raw_lines[raw_lines.len() - keep_lines..];
        let mut out = String::new();
        for line in kept {
            out.push_str(line);
            out.push('\n');
        }
        fs::write(&self.path, out)?;
        Ok(())
    }

    pub fn keep_line_count(&self, target_bytes: usize) -> Result<usize> {
        let lines = self.lines()?;
        if lines.is_empty() {
            return Ok(0);
        }
        let sizes: Vec<usize> = lines
            .iter()
            .map(|v| serde_json::to_string(v).map(|s| s.len() + 1))
            .collect::<std::result::Result<Vec<_>, _>>()?;
        let turn_start: Vec<bool> = lines
            .iter()
            .map(|v| {
                v.get("role").and_then(Value::as_str) == Some("user")
                    && v.get("content").is_some_and(Value::is_string)
            })
            .collect();

        let mut keep = 0usize;
        let mut bytes = 0usize;
        for i in (0..lines.len()).rev() {
            if keep > 0 && bytes + sizes[i] > target_bytes {
                break;
            }
            bytes += sizes[i];
            keep += 1;
        }
        if keep == 0 {
            keep = 1;
        }
        let mut start = lines.len() - keep;
        let mut adjusted = start;
        while adjusted < lines.len() && !turn_start[adjusted] {
            adjusted += 1;
        }
        if adjusted < lines.len() {
            start = adjusted;
        } else {
            while start > 0 && !turn_start[start] {
                start -= 1;
            }
        }
        Ok(lines.len() - start)
    }

    fn append_line(&self, value: &Value) -> Result<()> {
        let mut file = fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(&self.path)?;
        writeln!(file, "{}", serde_json::to_string(value)?)?;
        Ok(())
    }
}

pub fn first_line(s: &str) -> &str {
    s.lines().next().unwrap_or(s)
}

pub fn edit_diff_summary(path: &str, diff: &str) -> String {
    let mut added = 0usize;
    let mut removed = 0usize;
    for line in diff.lines() {
        if line.starts_with('+') && !line.starts_with("+++") {
            added += 1;
        }
        if line.starts_with('-') && !line.starts_with("---") {
            removed += 1;
        }
    }
    format!("Edit({path}) [+{added} -{removed} lines]")
}

fn line_count(s: &[u8]) -> usize {
    if s.is_empty() {
        0
    } else {
        s.iter().filter(|&&c| c == b'\n').count() + 1
    }
}

pub fn build_tool_call_summary(
    name: &str,
    fields: &std::collections::BTreeMap<String, String>,
) -> String {
    let mut label = String::new();
    match name {
        "Read" | "Write" | "Edit" => label = fields.get("path").cloned().unwrap_or_default(),
        "Glob" | "Grep" => label = fields.get("pattern").cloned().unwrap_or_default(),
        "Bash" => {
            label = fields
                .get("command")
                .cloned()
                .unwrap_or_default()
                .replace('\n', " ");
            if label.chars().count() > 80 {
                let truncated: String = label.chars().take(77).collect();
                label = format!("{truncated}...");
            }
        }
        "TodoWrite" => {
            if let Some(summary) = fields.get("summary").cloned() {
                if !summary.is_empty() {
                    label = summary;
                }
            }
            if label.is_empty() {
                // Compute progress from todos array
                if let Some(todos) = fields.get("todos") {
                    // todos is a JSON array string; parse it
                    if let Ok(arr) = serde_json::from_str::<Vec<Value>>(todos) {
                        let total = arr.len();
                        let completed = arr
                            .iter()
                            .filter(|item| {
                                item.get("status").and_then(Value::as_str) == Some("completed")
                            })
                            .count();
                        label = format!("{completed}/{total}");
                    }
                }
            }
        }
        "Skill" => label = fields.get("name").cloned().unwrap_or_default(),
        "SubAgent" => label = fields.get("description").cloned().unwrap_or_default(),
        _ => {}
    }
    if label.is_empty() {
        name.to_string()
    } else {
        format!("{name}({label})")
    }
}
